### PR TITLE
Add a discussion of sharding to the FFI tutorial

### PR DESCRIPTION
**Pull request by dfm**
Many users of the FFI want to be able to use their functions within parallel computations, and questions come up frequently enough that it makes sense to add this discussion to the introductory tutorial. Furthermore, this was the only remaining topic that was discussed in the [legacy custom call tutorial](https://jax.readthedocs.io/en/latest/Custom_Operation_for_GPUs.html), that hadn't been updated to the FFI case.

### CODE REVIEW

File: docs/ffi.md [hunk 1]
```diff
@@ -493,13 +501,179 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
...
+But, the story is a little bit more complicated for FFI calls.
+Since the internals of an FFI call are opaque to both JAX and XLA, FFI calls won't typically show optimal (or even good) performance when the data are sharded.
+
+Before getting into the FFI details, let's consider the behavior of our pure-JAX reference implementation of RMS normalization (the `rms_norm_ref` function defined at the top of this document) with a sharded input.
+As discussed above, our implementation treats all leading axes of the input as _batch_ dimensions, and the normalization is performed along the last axis.
+This means that if the data are sharded along any batch dimensions, but replicated on the last dimension, no communication is required.
+This can be seen by sharding our 2-dimensional test data from above along its first dimension and checking the compiled HLO for operations like `all-gather`, `all-reduce`, etc.:
+
+```{code-cell} ipython3
+import jax.sharding as shd
```
Comment (yashk2810) [inline docs/ffi.md:519]: Can we not import it like this? Let's keep it consistent like in other docs.

**Comment by dfm** [comment]
For reference, here is the rendered version of this section: https://jax--25771.org.readthedocs.build/en/25771/ffi.html#sharding

File: docs/ffi.md [hunk 2]
```diff
@@ -493,13 +501,179 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
...
+
+assert len(jax.devices()) == 4  # Set using the XLA_FLAGS environment variable
+mesh = jax.make_mesh((4,), ("x",))
+
+batch_shd = shd.NamedSharding(mesh, shd.PartitionSpec("x", None))
```
Comment (yashk2810) [inline docs/ffi.md:524]: Uae P instead of PartitionSpec

Comment (dfm) [reply docs/ffi.md:524]: If I truly must then okay lol :D

File: docs/ffi.md [hunk 3]
```diff
@@ -493,13 +501,174 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
...
+But, the story is a little bit more complicated for FFI calls.
+Since the internals of an FFI call are opaque to both JAX and XLA, FFI calls won't typically show optimal (or even good) performance when the data are sharded.
+
+Before getting into the FFI details, let's consider the behavior of our pure-JAX reference implementation of RMS normalization (the `rms_norm_ref` function defined at the top of this document) with a sharded input.
+As discussed above, our implementation treats all leading axes of the input as _batch_ dimensions, and the normalization is performed along the last axis.
+This means that if the data are sharded along any batch dimensions, but replicated on the last dimension, no communication is required.
+This can be seen by sharding our 2-dimensional test data from above along its first dimension and checking the compiled HLO for operations like `all-gather`, `all-reduce`, etc.:
+
+```{code-cell} ipython3
+from jax.sharding import NamedSharding, PartitionSpec as P
```
Comment (yashk2810) [inline docs/ffi.md:519]: You can use `NamedSharding` from the `jax` namespace. So `jax.NamedSharding`

File: docs/ffi.md [hunk 4]
```diff
@@ -493,13 +501,174 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
 
 and there will be no runtime overhead to using {func}`jax.lax.platform_dependent`, and the compiled program won't include any references to unavailable FFI targets.
 
++++
+
+## Sharding
+
+Most large scale users of JAX use its APIs for distributed computation across multiple devices.
+As discussed in {ref}`sharded-computation`, parallelism in JAX in controlled by sharding data across devices, and most JAX operations can be used within any of the supported parallel programming paradigms (from automatic to fully manual).
```
Comment (superbobry) [inline docs/ffi.md:509]: Typo: "is controlled"?

File: docs/ffi.md [hunk 5]
```diff
@@ -493,13 +501,174 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
...
+```{code-cell} ipython3
+hlo_data_shmap = jax.jit(rms_norm_shmap, out_shardings=data_shd).lower(x_data_shd).compile().as_text()
+assert "all-to-all" in hlo_data_shmap
+```
+
+### Using `custom partitioning`
+
+If you can't use {func}`~jax.experimental.shard_map.shard_map`, an alternative approach is to use {func}`~jax.experimental.custom_partitioning.custom_partitioning`, which supports automatic parallelization via {func}`jax.jit`.
+{func}`~jax.experimental.custom_partitioning.custom_partitioning` works by adding Python callbacks into the XLA compiler's partitioning pass, which allows very flexible logic, but also comes with some rough edges.
+We won't go into too much detail on the caveats here, but the main issues that you should be aware of are:
```
Comment (superbobry) [inline docs/ffi.md:592]: Should these be documented separately (apologies if they already are and I missed that)?

File: docs/ffi.md [hunk 6]
```diff
@@ -493,13 +501,174 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
...
+
+1. `custom_partitioning` can cause unexpected cache misses when used with the JAX's [Persistent compilation cache](https://jax.readthedocs.io/en/latest/persistent_compilation_cache.html). This can be mitigated using the `jax_remove_custom_partitioning_ptr_from_cache_key` configuration flag, but that isn't always appropriate either.
+2. Debugging `custom_partitioning` logic can be tedious because Python errors don't always get propagated, instead causing your Python process to exit.
```
Comment (superbobry) [inline docs/ffi.md:595]: Ooof, this is a very (very!) rough edge. Are they at least logged?

File: docs/ffi.md [hunk 7]
```diff
@@ -493,13 +501,174 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
...
+
+However, it's worth noting that the behavior is _different_ when the input is sharded along the data dimension.
+When used under `shard_map`, the data are resharded on the batch dimension, whereas with `custom_partitioning` the data are gathered onto each device.
+
+```{code-cell} ipython3
+hlo_data_partitioned = jax.jit(rms_norm_partitioned, out_shardings=data_shd).lower(x_data_shd).compile().as_text().strip()
+assert "all-gather" in hlo_data_partitioned
+```
+
+To also support automatic parallelization of the backwards pass, we would also need to write (similar) {func}`~jax.experimental.custom_partitioning.custom_partitioning` rules for `rms_norm_fwd` and `rms_norm_bwd`, but we leave those as an exercise for the reader.
```
Comment (superbobry) [inline docs/ffi.md:663]: > but we leave those as an exercise for the reader.

😆

**Review by superbobry — APPROVED** [review]

File: docs/ffi.md (hunk 3 above)
Comment (dfm) [reply docs/ffi.md]: Done!

File: docs/ffi.md (hunk 6 above)
Comment (dfm) [reply docs/ffi.md]: Good point! Yes, they do get logged, and I've added a comment to that effect.

The issue arises here:

https://github.com/openxla/xla/blob/9068ccbdcb10b0d143fea6d1296a6084b6d0cb43/xla/python/custom_partition_callback.cc#L182

because the GSPMD partitioner doesn't have any mechanisms for indicating errors during sharding _propagation_. I think that exceptions during _partitioning_ should fail gracefully.

File: docs/ffi.md (hunk 5 above)
Comment (dfm) [reply docs/ffi.md:592]: Good question! No, I don't think so. I also don't think this is necessarily the right place to document all of this, but I thought it was important to comment on these specific issues because I hit them when writing up this tutorial. I'm inclined to leave this as is, and hopefully things will get a little cleaner with the shardy approach in the future, or something else!

File: docs/ffi.md (hunk 1 above)
Comment (dfm) [reply docs/ffi.md]: Done.

**Review by pschuh — APPROVED** [review]
